Modernize MaterialX version checks for easy reading and streamlined logic

scotbrew opened the pull request:
### Description of Change(s)
Use the new MATERIALX_VERSION_INDEX and MATERIALX_GENERATE_INDEX added in MaterialX 1.39.2.  This makes the version check code easier to read and streamlines the logic to avoids potential version check errors when using MATERIALX_MAJOR_VERSION, MATERIALX_MINOR_VERSION, and MATERIALX_BUILD_VERSION comparisons directly.

**Sample usage:**
`#if MATERIALX_VERSION_INDEX < MATERIALX_GENERATE_INDEX(1, 39, 0)`

Note: MATERIALX_VERSION_INDEX and MATERIALX_GENERATE_INDEX were added as of MaterialX 1.39.2 and then revised in 1.39.4.  Defining these macros if MaterialX version < 1.39.4.

### Fixes Issue(s)
https://github.com/PixarAnimationStudios/OpenUSD/issues/3651

### Testing
Manually tested compilation with MaterialX 1.39.3 and 1.38.10.

### Checklist

- [x] I have created this PR based on the dev branch

- [x] I have followed the [coding conventions](https://openusd.org/release/api/_page__coding__guidelines.html)

- [x] I have added unit tests that exercise this functionality (Reference: 
  [testing guidelines](https://openusd.org/release/api/_page__testing__guidelines.html))

- [x] I have verified that all unit tests pass with the proposed changes

- [x] I have submitted a signed Contributor License Agreement (Reference: 
  [Contributor License Agreement instructions](https://openusd.org/release/contributing_to_usd.html#contributor-license-agreement))

scotbrew committed:
Modernize MaterialX version checks

Use new MATERIALX_VERSION_INDEX and MATERIALX_GENERATE_INDEX added in MaterialX 1.39.2.  This makes the version check code easier to read and avoids potential version check errors when using MATERIALX_MAJOR_VERSION,
MATERIALX_MINOR_VERSION, and MATERIALX_BUILD_VERSION comparisons directly.

Sample usage:
#if MATERIALX_VERSION_INDEX < MATERIALX_GENERATE_INDEX(1, 39, 0)

MATERIALX_VERSION_INDEX and MATERIALX_GENERATE_INDEX were added as of MaterialX 1.39.2 and then revised in 1.39.4.  Will define these macros if version < 1.39.4.

Tested compilation with MaterialX 1.39.3 and 1.38.10.

jesschimein commented:
Filed as internal issue #USD-11055

(This is an automated message. See [here](https://openusd.org/release/contributing_to_usd.html#github-issues) for more information.)

scotbrew committed:
Move MATERIALX_GENERATE_INDEX macro

Moving MATERIALX_GENERATE_INDEX and MATERIALX_VERSION_INDEX above the MaterialX ShaderGenerator includes as there will likely be another include conditionally added for WGSL based on the MaterialX version.  Putting the code in the same place so it does not have to be moved later and create merging confusion.

Tested with MaterialX 1.39.3 and 1.38.10 versions on MacOS.

scotbrew commented on pxr/imaging/hdSt/materialXShaderGen.h:
```diff
@@ -8,11 +8,22 @@
 #define PXR_IMAGING_HD_ST_MATERIALX_SHADER_GEN_H
 
 #include "pxr/pxr.h"
+#include <MaterialXCore/Library.h>
```
Added this `#include <MaterialXCore/Library.h>` (which declares the MaterialX version defines) and moved the defines above the other MaterialX includes as there are plans in the future for conditionally importing MaterialXGenMsl/WgslShaderGenerator.h based on the MaterialX version.
